### ISSUE TITLE
GetNextOccurrence throws if next occurrence produces invalid time

### ISSUE THREAD

**Issue by atifaziz** [atifaziz/NCrontab#21]
## Step to Reproduce

Install [NCrontab 3.2.0](https://www.nuget.org/packages/ncrontab/3.2.0) and then run the following C# script assuming `NCrontab.dll` is in the same folder as the script:

```c#
#r "NCrontab.dll"
using NCrontab;
var schedule = CrontabSchedule.Parse("0 0 29 Feb Mon");
var occurrences = schedule.GetNextOccurrences(new DateTime(9000, 1, 1), DateTime.MaxValue);
foreach (var dt in occurrences)
    Console.WriteLine(dt.ToString("ddd, dd, MMM yyyy"));
```

## Expected Output

Expected the sequence of occurrences to end with the date Feb 29, 9988.

## Actual Output

The sequence throws `ArgumentOutOfRangeException` instead of ending after yielding Feb 29, 9988:

```
Mon, 29, Feb 9008
Mon, 29, Feb 9036
Mon, 29, Feb 9064
Mon, 29, Feb 9092
Mon, 29, Feb 9104
Mon, 29, Feb 9132
Mon, 29, Feb 9160
Mon, 29, Feb 9188
Mon, 29, Feb 9216
Mon, 29, Feb 9244
Mon, 29, Feb 9272
Mon, 29, Feb 9312
Mon, 29, Feb 9340
Mon, 29, Feb 9368
Mon, 29, Feb 9396
Mon, 29, Feb 9408
Mon, 29, Feb 9436
Mon, 29, Feb 9464
Mon, 29, Feb 9492
Mon, 29, Feb 9504
Mon, 29, Feb 9532
Mon, 29, Feb 9560
Mon, 29, Feb 9588
Mon, 29, Feb 9616
Mon, 29, Feb 9644
Mon, 29, Feb 9672
Mon, 29, Feb 9712
Mon, 29, Feb 9740
Mon, 29, Feb 9768
Mon, 29, Feb 9796
Mon, 29, Feb 9808
Mon, 29, Feb 9836
Mon, 29, Feb 9864
Mon, 29, Feb 9892
Mon, 29, Feb 9904
Mon, 29, Feb 9932
Mon, 29, Feb 9960
Mon, 29, Feb 9988
System.AggregateException: One or more errors occurred. ---> System.ArgumentOutOfRangeException: Valid values are between 1 and 9999, inclusive.
Parameter name: year
   at System.Globalization.GregorianCalendar.GetDaysInMonth(Int32 year, Int32 month, Int32 era)
   at NCrontab.CrontabSchedule.GetNextOccurrence(DateTime baseTime, DateTime endTime)
   at NCrontab.CrontabSchedule.GetNextOccurrence(DateTime baseTime, DateTime endTime)
   at NCrontab.CrontabSchedule.GetNextOccurrence(DateTime baseTime, DateTime endTime)
   at NCrontab.CrontabSchedule.<GetNextOccurrences>d__15.MoveNext()
   at Submission#0.<<Initialize>>d__0.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.CodeAnalysis.Scripting.ScriptExecutionState.<RunSubmissionsAsync>d__9`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.CodeAnalysis.Scripting.Script`1.<RunSubmissionsAsync>d__21.MoveNext()
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification)
   at Microsoft.CodeAnalysis.Scripting.Hosting.CommandLineRunner.RunScript(ScriptOptions options, String code, ErrorLogger errorLogger, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Scripting.Hosting.CommandLineRunner.RunInteractiveCore(ErrorLogger errorLogger)
   at Microsoft.CodeAnalysis.Scripting.Hosting.CommandLineRunner.RunInteractive()
   at Microsoft.CodeAnalysis.CSharp.Scripting.Hosting.Csi.Main(String[] args)
---> (Inner Exception #0) System.ArgumentOutOfRangeException: Valid values are between 1 and 9999, inclusive.
Parameter name: year
   at System.Globalization.GregorianCalendar.GetDaysInMonth(Int32 year, Int32 month, Int32 era)
   at NCrontab.CrontabSchedule.GetNextOccurrence(DateTime baseTime, DateTime endTime)
   at NCrontab.CrontabSchedule.GetNextOccurrence(DateTime baseTime, DateTime endTime)
   at NCrontab.CrontabSchedule.GetNextOccurrence(DateTime baseTime, DateTime endTime)
   at NCrontab.CrontabSchedule.<GetNextOccurrences>d__15.MoveNext()
   at Submission#0.<<Initialize>>d__0.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.CodeAnalysis.Scripting.ScriptExecutionState.<RunSubmissionsAsync>d__9`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.CodeAnalysis.Scripting.Script`1.<RunSubmissionsAsync>d__21.MoveNext()<---
```

**Comment by Styxxy** [comment]
Internally the GetNextOccurrences method calls the GetNextOccurrence method. This method always returns a DateTime, which cannot be null.

I found in this case also another "problem". If for above schedule, you request a next occurrence with start/end date that would not generate any occurrence for the given schedule; the end date will be returned.

Sample:

    var schedule = CrontabSchedule.Parse("0 0 29 Feb Mon");
    var occurrence = schedule.GetNextOccurrence(new DateTime(2017, 1, 1), new DateTime(2017, 12, 31));
    // occurrence = "2017-12-31"

One option to fix this issue (#21) : the method GetNextOccurrence can be extracted to a private method that returns a nullable DateTime value. The public method would still return the end date if the private method returns null (or else the value). The GetNextOccurrences method would stop returning.

Would you agree with above approach?

**Comment by atifaziz** [comment]
> Would you agree with above approach?

@Styxxy Sounds reasonable.
